clients/da-vinci-client/src/main/java/com/linkedin/davinci/client/DaVinciRecordTransformer.java

```diff
@@ -0,0 +1,47 @@
+package com.linkedin.davinci.client;
+
+import com.linkedin.venice.utils.lazy.Lazy;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericRecord;
+
+
+/**
+ * This interface can be implemented in order to transform records stored in the Da Vinci Client.
+ *
+ * The input is what is consumed from the raw Venice data set, whereas the output is what is stored
+ * into Da Vinci's local storage (e.g. RocksDB).
+ *
+ * N.B.: The inputs are wrapped inside {@link Lazy} so that if the implementation need not look at
+ * them, the deserialization cost is not paid.
+ *
+ * @param <K> type of the input key
+ * @param <V> type of the input value
+ * @param <O> type of the output value
+ */
+public interface DaVinciRecordTransformer<K, V, O extends GenericRecord> {
```

Comment:
Regarding O extending GenericRecord. This might be too loose.  It seems from the code we require that GenericRecord have "key" and "value", so the structure of the output isn't really up to the user to define.

We probably need some kind of Record class to contain the outputted K/V in order to be sure the code will work (otherwise if the user just renders some random GenericRecord without key or value fields, it'll NPE).